### PR TITLE
Switch caching package's interface to be generic and add experimental flag to try different caches

### LINKED ISSUE #1979: Evaluate replacing ristretto with theine-go
### Problem Statement

Ristretto is non-ideal for a bunch of reasons:

- Ristretto is somewhat unmaintained; there are some ex-dgraph folks under the organization outcaste-io triaging but not actively developing.
- Major users like Vitess have replaced it.
- I've now seen both the [theine](https://github.com/Yiling-J/theine-go/blob/v0.3.2/README.md#hit-ratios) and [otto](https://github.com/maypok86/otter/blob/v1.2.1/README.md#performance) developers very much so struggle to reproduce ristretto's benchmarks leading me to believe our usage could also be misconfigured leading to significantly worse performance. [This comment](https://github.com/dgraph-io/ristretto/issues/363#issuecomment-2120183298) recently caught my attention
- Ristretto doesn't support Go's generics
- Ristretto doesn't implement all of the things that make Caffeine (the premier java tinylfu implementation) perform optimally

### Solution Brainstorm

Support [theine](https://github.com/Yiling-J/theine-go) or [otto](https://github.com/maypok86/otter) using our cache interface and a hidden feature flag until we can confirm performance and deprecate ristretto

### PR DESCRIPTION
**Pull request by josephschorr**
Fixes #1979

### CODE REVIEW

**Review by vroldanbet — DISMISSED** [review]
Changes LGTM

File: pkg/cache/cache_theine.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Yiling-J/theine-go"
+	"github.com/rs/zerolog"
+)
+
+func NewTheineCacheWithMetrics[K KeyString, V any](name string, config *Config) (Cache[K, V], error) {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply pkg/cache/cache_theine.go:14]: Otter's own README shows it is less performant for DB-like services. We could always add it as a followup

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -7,6 +7,19 @@ import (
 	"github.com/rs/zerolog"
 )
 
+// KeyString is an interface for keys that can be converted to strings.
+type KeyString interface {
+	comparable
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply pkg/cache/cache.go:12]: How would we use a `[]byte` without wrapping it or casting?

File: pkg/cache/cache_theine.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		return key.KeyString()
+	})
+	built, err := builder.Build()
+	if err != nil {
+		return nil, err
+	}
+	return &wrappedTC[K, V]{built, config.DefaultTTL, sync.Once{}}, nil
+}
+
+type wrappedTC[K KeyString, V any] struct {
```
[the remark this answers is not in the document]
Comment (josephschorr) [reply pkg/cache/cache_theine.go]: Updated

File: pkg/cache/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/cache/cache.go:12]:
```go
type CaveatNameFromProto []byte

func (c CaveatNameFromProto) KeyString() string {
  return string(c)
}
```

[the remark this answers is not in the document]
Comment (josephschorr) [reply pkg/cache/cache.go:12]: Yeah, wrapped. I kept the `comparable` for the Ristretto cache impl.

File: pkg/cache/cache_theine.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jzelinskie) [reply pkg/cache/cache_theine.go:14]: let's add both and blog about it

**Review by ecordell — DISMISSED** [review]

**Review by ecordell — APPROVED** [review]

**Pull request merged**